common/statistics/stateStatistics.go

```diff
@@ -0,0 +1,225 @@
+package statistics
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+)
+
+type stateStatistics struct {
+	numCache         uint64
+	numSyncCache     uint64
+	numSnapshotCache uint64
+
+	numPersister         sync.Map
+	numSyncPersister     sync.Map
+	numSnapshotPersister sync.Map
+
+	numTrie uint64
+}
+
+// NewStateStatistics returns a structure able to collect statistics for state
+func NewStateStatistics() *stateStatistics {
+	return &stateStatistics{}
+}
+
+func (ss *stateStatistics) resetMap(counters *sync.Map) {
+	counters.Range(func(k, v interface{}) bool {
+		counters.Store(k, uint64(0))
+		return true
+	})
+}
+
+func (ss *stateStatistics) incrMapOp(counters *sync.Map, epoch uint32) {
+	val, _ := counters.LoadOrStore(epoch, uint64(0))
+	counters.Store(epoch, val.(uint64)+uint64(1))
+}
+
+func (ss *stateStatistics) getMapOp(counters *sync.Map, epoch uint32) uint64 {
+	val, ok := counters.Load(epoch)
+	if !ok {
+		return uint64(0)
+	}
+
+	return val.(uint64)
+}
+
+// ResetAll will reset all statistics
+func (ss *stateStatistics) ResetAll() {
+	ss.Reset()
+	ss.ResetSync()
+	ss.ResetSnapshot()
+}
+
+// Reset will reset processing statistics
+func (ss *stateStatistics) Reset() {
+	atomic.StoreUint64(&ss.numCache, 0)
+
+	ss.resetMap(&ss.numPersister)
+
+	atomic.StoreUint64(&ss.numTrie, 0)
+}
+
+// ResetSync will set sync statistics based on processing statistics
+func (ss *stateStatistics) ResetSync() {
```

Comment:
That way `getSyncStats()` will also simplify a lot

Reply:
Since i don't have an easy way to differenciate between normal processing and sync processing, and since no other processing is done when sync is in progress, i get sync statistics based on normal processing. This way, when the sync is starting, the sync stats will be set to the (old) current value on normal processing stats, and when the sync is finshed, the sync stats will be calculated based on (new) current value for normal processing and the already set sync stats values

Reply:
I see. In this case remove the methods that are not needed: `IncrSyncCache()` , `SyncCache()` , `IncrSyncPersister()`, etc.